### PR TITLE
tests: support snap directory in external tests

### PR DESCRIPTION
**Pull request by elopio**
LP: [#1671620](https://bugs.launchpad.net/snapcraft/+bug/1671620)

### CODE REVIEW

File: external_snaps_tests/__main__.py [hunk 1]
```diff
@@ -97,9 +97,13 @@ def _bzr_branch(url):
...
     try:
-        for dirpath, _, filenames in os.walk(path):
-            if 'snapcraft.yaml' in filenames or '.snapcraft.yaml' in filenames:
+        for dirpath, dirnames, filenames in os.walk(path, topdown=True):
+            if ('snap' in dirnames
+                    or 'snapcraft.yaml' in filenames
+                    or '.snapcraft.yaml' in filenames):
                 _build_snap(dirpath, cleanbuild, keep_dir)
+                # Do not recurse in any directory.
+                dirnames = []
```
Comment (kyrofa) [inline external_snaps_tests/__main__.py]: I've never seen this before. It prevents the `walk` from going into those directories?

Comment (elopio) [inline external_snaps_tests/__main__.py]: Just if you use the top down argument.

**Comment by elopio** [comment]
I'm testing this with a run of the nightlies, so wait a moment before landing.

File: external_snaps_tests/__main__.py (hunk 1 above)
Comment (kyrofa) [inline external_snaps_tests/__main__.py]: @elopio but you're not actually modifying the list in-place, right?

> When topdown is True, the caller can modify the dirnames list in-place (perhaps using del or slice assignment), and walk() will only recurse into the subdirectories whose names remain in dirnames[...]

Comment (elopio) [inline external_snaps_tests/__main__.py]: I am. I think. I don't really understand your question.
The docs say that recursion will happen only on the directories that remain in dirnames. If dirnames is empty, no recursion will happen.

**Comment by elopio** [comment]
Hold this. There's a directory called snap in etcd, that's not related to us. This breaks the docker test.

File: external_snaps_tests/__main__.py (hunk 1 above)
Comment (kyrofa) [inline external_snaps_tests/__main__.py]: > I am. I think. I don't really understand your question.

Let me explain with this:

```python
>>> real = [1, 2, 3]
>>> ref = real
>>> real
[1, 2, 3]
>>> ref
[1, 2, 3]
>>> ref = []  # you're doing this
>>> real
[1, 2, 3]
>>> ref
[]
>>> ref = real
>>> del ref[:]
>>> real
[]
>>> ref
[]
```

Comment (elopio) [inline external_snaps_tests/__main__.py]: You are right, thanks Kyle.
And with this, the change for etcd wouldn't have been necessary. But it doesn't hurt either.

**Comment by elopio** [comment]
Ready!

**Pull request merged**